### ISSUE TITLE
Hyperparameter search

### ISSUE THREAD

**Issue by alecgunny** [ML4GW/aframe#36]
Use [Ray Tune](https://pytorch.org/tutorials/beginner/hyperparameter_tuning_tutorial.html) to perform hyperparameter searching on top of the training function built by #28 . Obviously learning rate will be the first and most critical, but once the training function is built we can decide what else should go in there.

**Comment by alecgunny** [comment]
Re-upping this now that we're moving towards publication. While having something robust and parallelized will be good longer-term, in the short term even a manual serial search will be useful for ensuring that we're in the right ballpark. It will also be important for having an explanation of _how_ we arrived at our hyperparameters in any paper. Most important hyperparameters for now are probably:
- `learning_rate`
- `lr_ramp_epochs`
- `weight_decay`
- `glitch_prob`
- `waveform_prob`
- Not in `main` yet, but downweighting of probability of injecting signals on top of glitches

**Comment by alecgunny** [comment]
Worth keeping an eye on [this project](https://github.com/ML4GW/APwML-Hackathon-HP-Search) for a my robust longer term solution

**Comment by wbenoit26** [comment]
Issues #271, #273, and #274 should be resolved before this issue is worked - any others?

**Comment by alecgunny** [comment]
@wbenoit26 I believe #255 and #256 should be resolved first too

**Comment by wbenoit26** [comment]
Completed a hyperparameter search on March 7th, 2023 over the following parameter space with 64 iterations:

```
hp_set["waveform_prob"] = rng.uniform(0.1, 0.9, n_iter)
# Capping glitch prob due to imbalance of glitch dataset
hp_set["glitch_prob"] = rng.uniform(0.1, 0.65, n_iter)
hp_set["max_lr"] = np.exp(rng.uniform(-5, -1, n_iter))
hp_set["lr_ramp_epochs"] = rng.integers(50, 100, n_iter)
hp_set["weight_decay"] = np.exp(rng.uniform(-6, -4, n_iter))

# Ensure that waveform_prop doesn't get boosted above 1
min_dw = 1 - (1 - np.sqrt(hp_set["waveform_prob"])) / hp_set["glitch_prob"]
hp_set["glitch_downweight"] = [rng.uniform(max(dw, 0), 1) for dw in min_dw]
```

There were three models that had perfect performance on the validation set. These models had the following parameter sets:

Iteration 10:
```
waveform_prob: 0.450466
glitch_prob: 0.535603
max_lr: 0.017585
lr_ramp_epochs: 61
weight_decay: 1.17965e-05
glitch_downweight 0.970870
```

Iteration 21:
```
waveform_prob: 0.462856
glitch_prob: 0.306851
max_lr: 0.153796
lr_ramp_epochs: 50
weight_decay: 1.79051e-05
glitch_downweight: 0.865457
```

Iteration 60:
```
waveform_prob: 0.661827, 
glitch_prob: 0.593902, 
max_lr: 0.151186, 
lr_ramp_epochs: 85, 
weight_decay: 1.06096e-06, 
glitch_downweight: 0.899555
```

The models trained during this search can be found on LHO here: `/home/william.benoit/hp_search_results_20230307`
